include/boost/numeric/ublas/tensor/detail/extents_functions.hpp

```diff
@@ -0,0 +1,271 @@
+    return a > typename ExtentsType::value_type(1);
+  };
+  auto equal_one = [](auto const &a) { return a == typename ExtentsType::value_type(1); };
+
+  if constexpr (!detail::is_static_extents<ExtentsType>::value) {
+    return std::any_of(e.begin(), e.begin() + 2, greater_one) &&
+           std::any_of(e.begin(), e.begin() + 2, equal_one) &&
+           std::all_of(e.begin() + 2, e.end(), equal_one);
+  } else {
+    auto arr = e.to_array();
+    return std::any_of(arr.begin(), arr.begin() + 2, greater_one) &&
+           std::any_of(arr.begin(), arr.begin() + 2, equal_one) &&
+           std::all_of(arr.begin() + 2, arr.end(), equal_one);
+  }
+}
+
+/** @brief Returns true if this has a matrix shape
+ *
+ * @returns true if (m,n,[1,...,1]) with m > 1 and n > 1
+ */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+BOOST_UBLAS_TENSOR_INLINE 
+constexpr bool is_matrix(ExtentsType const &e) {
+  if (e.size() < typename ExtentsType::size_type(2)) {
+    return false;
+  }
+
+  auto greater_one = [](auto const &a) {
+    return a > typename ExtentsType::value_type(1);
+  };
+  auto equal_one = [](auto const &a) { return a == typename ExtentsType::value_type(1); };
+
+  if constexpr (!detail::is_static_extents<ExtentsType>::value) {
+    return std::all_of(e.begin(), e.begin() + 2, greater_one) &&
+           std::all_of(e.begin() + 2, e.end(), equal_one);
+  } else {
+    auto arr = e.to_array();
+    return std::all_of(arr.begin(), arr.begin() + 2, greater_one) &&
+           std::all_of(arr.begin() + 2, arr.end(), equal_one);
+  }
+}
+
+/** @brief Returns true if this is has a tensor shape
+ *
+ * @returns true if !empty() && !is_scalar() && !is_vector() && !is_matrix()
+ */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+BOOST_UBLAS_TENSOR_INLINE 
+constexpr bool is_tensor(ExtentsType const &e) {
+  if (e.size() < typename ExtentsType::size_type(3)) {
+    return false;
+  }
+
+  auto greater_one = [](auto const &a) {
+    return a > typename ExtentsType::value_type(1);
+  };
+
+  if constexpr (!detail::is_static_extents<ExtentsType>::value) {
+    return std::any_of(e.begin() + 2, e.end(), greater_one);
+  } else {
+    auto arr = e.to_array();
+    return std::any_of(arr.begin() + 2, arr.end(), greater_one);
+  }
+}
+
+/** @brief Eliminates singleton dimensions when size > 2
+ *
+ * squeeze {  1,1} -> {  1,1}
+ * squeeze {  2,1} -> {  2,1}
+ * squeeze {  1,2} -> {  1,2}
+ *
+ * squeeze {1,2,3} -> {  2,3}
+ * squeeze {2,1,3} -> {  2,3}
+ * squeeze {1,3,1} -> {  1,3}
+ *
+ * @returns basic_extents<int_type> with squeezed extents
+ */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+auto squeeze(ExtentsType const &e) {
+  
+  if (e.size() <= 2) {
+    if constexpr (detail::is_static_extents<ExtentsType>::value) {
+      return e.to_dynamic_extents();
```

Comment:
created the compile-time squeeze and removed `to_dynamic_extents` from static_extents but static rank with dynamic extents cannot be used with it because of runtime constraints we do not the how much in size it will decrease.